syz-verifier/main.go

```diff
@@ -0,0 +1,310 @@
+	}
+	execBin := cfg.ExecutorBin
+	if !osutil.IsExist(execBin) {
+		log.Fatalf("bad syzkaller config: can't find %v", execBin)
+	}
+
+	crashdir := filepath.Join(workdir, "crashes")
+	osutil.MkdirAll(crashdir)
+	for idx := range pools {
+		OS, Arch := target.OS, target.Arch
+		targetPath := OS + "-" + Arch + "-" + strconv.Itoa(idx)
+		osutil.MkdirAll(filepath.Join(workdir, targetPath))
+		osutil.MkdirAll(filepath.Join(crashdir, targetPath))
+	}
+
+	for idx, pi := range pools {
+		var err error
+		pi.Reporter, err = report.NewReporter(pi.cfg)
+		if err != nil {
+			log.Fatalf("failed to create reporter for instance-%d: %v", idx, err)
+		}
+		pi.vmRunners = make(map[int][]int)
+		pi.progs = make([]int, 0)
+	}
+
+	calls := make(map[*prog.Syscall]bool)
+	for _, id := range cfg.Syscalls {
+		calls[target.Syscalls[id]] = true
+	}
+
+	vrf := &Verifier{
+		workdir:     workdir,
+		crashdir:    crashdir,
+		pools:       pools,
+		target:      target,
+		choiceTable: target.BuildChoiceTable(nil, calls),
+		rnd:         rand.New(rand.NewSource(time.Now().UnixNano() + 1e12)),
+		runnerBin:   runnerBin,
+		executorBin: execBin,
+		addr:        addr,
+	}
+
+	srv, err := startRPCServer(vrf)
+	if err != nil {
+		log.Fatalf("failed to initialise RPC server: %v", err)
+	}
+
+	for idx, pi := range pools {
+		go func(pi *poolInfo, idx int) {
+			for {
+				// TODO: implement support for multiple VMs per Pool.
+				inst, err := pi.pool.Create(0)
+				if err != nil {
+					log.Fatalf("failed to create instance: %v", err)
+				}
+				defer inst.Close()
+
+				fwdAddr, err := inst.Forward(srv.port)
+				if err != nil {
+					log.Fatalf("failed to set up port forwarding: %v", err)
+
+				}
+
+				runnerBin, err := inst.Copy(vrf.runnerBin)
+				if err != nil {
+					log.Fatalf(" failed to copy runner binary: %v", err)
+				}
+				_, err = inst.Copy(vrf.executorBin)
+				if err != nil {
+					log.Fatalf("failed to copy executor binary: %v", err)
+				}
+
+				cmd := instance.RunnerCmd(runnerBin, fwdAddr, vrf.target.OS, vrf.target.Arch, idx, 0)
+				outc, errc, err := inst.Run(pi.cfg.Timeouts.VMRunningTime, vrf.vmStop, cmd)
+				if err != nil {
+					log.Fatalf("failed to start runner: %v", err)
+				}
+
+				inst.MonitorExecution(outc, errc, pi.Reporter, vm.ExitTimeout)
+				srv.cleanup(idx, 0)
+			}
+		}(pi, idx)
+	}
+
+	select {}
+}
+
+func startRPCServer(vrf *Verifier) (*RPCServer, error) {
+	srv := &RPCServer{
+		vrf:   vrf,
+		pools: vrf.pools,
+		progs: make(map[int]*programInfo),
+	}
+
+	s, err := rpctype.NewRPCServer(vrf.addr, "Verifier", srv)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("serving rpc on tcp://%v", s.Addr())
+	srv.port = s.Addr().(*net.TCPAddr).Port
+
+	go s.Serve()
+	return srv, nil
+}
+
+// Connect notifies the RPCServer that a new Runner was started.
+func (srv *RPCServer) Connect(a *rpctype.RunnerConnectArgs, r *int) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	pool, vm := a.Pool, a.VM
+	srv.pools[pool].vmRunners[vm] = []int{}
```

Comment:
You can just assign nil.
But the question is: how can it be non-empty already at this point? And if it's anyhow non-empty at this point, we are dropping these programs at the floor and they will leak memory. I think you need to call cleanup method here as well.

Reply:
> This line is for creating an new entry in the map for this runner.

You don't need to do this. If you query a map for a non-existent element, it will return you the zero value, which for slices mean empty slice. So you don't need to create it to set to empty, it's already empty.

However, it may be useful then to add an assert to code that it's indeed empty. It's easy to get concurrency issues here with future modifications, or maybe it already can be non-empty in some tricky even ordering. E.g. we can get Connect from a crashed VM _after_ we get notification about the crash. So that delayed Connect can arrive even after Connect from a new runner with the same index.